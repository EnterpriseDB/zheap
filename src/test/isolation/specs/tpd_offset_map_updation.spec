# If the previous transaction slot points to a TPD slot then we need to update
# the slot in the offset map of the TPD entry
#

setup
{
	CREATE TABLE t (a int, b int);
	INSERT INTO t SELECT g,g FROM generate_series(1,1000)g;
}

teardown
{
	DROP TABLE t;
}

session "s1"
step "s1a" { BEGIN; }
step "s1b" { INSERT INTO t VALUES (1001); }
step "s1c" { SELECT * FROM t WHERE a = 1; }
step "s1d" { END; }

session "s2"
step "s2a" { BEGIN; }
step "s2b" { UPDATE t SET b = b + 1 WHERE a = 1; }
step "s2c" { ROLLBACK; }

session "s3"
step "s3a" { BEGIN; }
step "s3b" { UPDATE t SET b = b + 1 WHERE a = 2; }
step "s3d" { END; }

session "s4"
step "s4a" { BEGIN; }
step "s4b" { UPDATE t SET b = b + 1 WHERE a = 3; }
step "s4c" { END; }

permutation "s1a" "s1b" "s2b"  "s2a" "s2b" "s3a" "s3b" "s4a" "s4b" "s2c" "s1c"
"s1d" "s3d" "s4c"
