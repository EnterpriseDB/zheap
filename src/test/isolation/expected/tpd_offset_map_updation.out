Parsed test spec with 4 sessions

starting permutation: s1a s1b s2b s2a s2b s3a s3b s4a s4b s2c s1c s1d s3d s4c
step s1a: BEGIN;
step s1b: INSERT INTO t VALUES (1001);
step s2b: UPDATE t SET b = b + 1 WHERE a = 1;
step s2a: BEGIN;
step s2b: UPDATE t SET b = b + 1 WHERE a = 1;
step s3a: BEGIN;
step s3b: UPDATE t SET b = b + 1 WHERE a = 2;
step s4a: BEGIN;
step s4b: UPDATE t SET b = b + 1 WHERE a = 3;
step s2c: ROLLBACK;
step s1c: SELECT * FROM t WHERE a = 1;
a              b              

1              2              
step s1d: END;
step s3d: END;
step s4c: END;
